workspaces/ui-v2/src/optic-components/hooks/useShapeDescriptor.ts

```diff
@@ -49,6 +54,30 @@ const changesSinceShapeQuery = `
     }
 }`;
 
+type ShapeChoice =
+  | {
+      id: string;
+      jsonType: JsonLike.OBJECT;
+      asArray?: undefined;
+      asObject: IObjectRender;
+    }
+  | {
+      id: string;
+      jsonType: JsonLike.ARRAY;
+      asArray: IArrayRender;
+      asObject?: undefined;
+    }
+  | {
+      id: string;
+      jsonType: Exclude<JsonLike, JsonLike.OBJECT | JsonLike.ARRAY>;
```

Comment:
gonna have to take your typescript course @niclim :)

Reply:
Haha it's a lot less glamorous than the code - more of a try stuff and see if the compiler accepts my typings